File: client/src/game/utils/abilities.ts
export type AbilityFlag = number;

export const MULTISHOT_ABILITY_FLAG: AbilityFlag = 1 << 1;
export const WIDE_BEAM_ABILITY_FLAG: AbilityFlag = 1 << 2;

export function isAbilityActive(flags: AbilityFlag, abilityFlag: AbilityFlag): boolean {
  return (flags & abilityFlag) != 0;
}

File: server/internal/game/entities/abilities.go
package entities

import (
	"math/rand"
)

// An AbilityFlag is a number where each bit corresponds to a particular
// powerup ability. Multiple abilities can be active at the same time.
type AbilityFlag uint32

const (
	MultishotAbilityFlag AbilityFlag = 1 << 1 // shoots 3 bullets in parallel
	WideBeamAbilityFlag  AbilityFlag = 1 << 2 // shoots 1 bullet
	ShieldAbilityFlag    AbilityFlag = 1 << 3 // protects against 1 collision
)

func isAbilityActive(flags AbilityFlag, ability AbilityFlag) bool {
	return (flags & ability) != 0
}

func newRandomAbility() AbilityFlag {
	return AbilityFlag(1 << (1 + rand.Intn(3)))
}
